feat(cli) Add generateOptions to configuration

egodsk opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Currently you must specify `--no-spec` on *ALL* generate commands, to not pass the `--spec` flag onwards to the collection, that generates the files. No configuration is available to allow for the `--no-spec` command to be default behavior, unless otherwise specified.

This PR fixes/closes:
- #314 

## What is the new behavior?
With this feature, you are able to specify a `generateOptions` in the configuration file, which currently only have the ability to change the generation options for `spec`, however you can then specify `spec` to be either a boolean (true/false) or an object, which the different generate names as key, with booleans as value, like so:
```
"generateOptions": {
  "spec": true
}
```
or
```
"generateOptions": {
  "spec": {
    "service": true
    ...
  }
}
```

The arguments for generate has been slightly altered, to allow for this change. There has been added a new `--spec <true/false>` option, such that the `spec` parameter is not default to true when the command is passed to the action, but undefined.

This feature follows a hierarchy, such that the CLI arguments have highest priority, then comes the projects individual options (options specified within the `"projects": {}...` option), and lastly the overall `generateOptions` for the project.

Highest:
```
nest g service Cats --spec true
```

Lower:
```
{
  "sourceRoot": "src",
  "projects": {
    "cat-project": {
      "generateOptions": {
        "spec": true
      }
    }
  }
  ...
}
```

Lowest:
```
{
  "sourceRoot": "src",
  "generateOptions: {
   "spec": true
  }
  ...
}
```

Such that if `--spec true` is set on the CLI and the `spec` in the `nest-cli.json` options is set to false, it would overwrite and pass true to the action, and so forward.

One thing I’m unable to currently figure out, is that you can only use the generateOptions with the spec as an object if you use the full name of the schematic you want to run. Ex `service` and not `s` - if anyone has a good suggestion on how to be able to handle aliases, please let me know! :) 

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
[x] Maybe?
```
The generate command, has been changed such that the `spec` argument is no longer default to true, but undefined, unless otherwise specified in the CLI arguments. However for the `nestjs/schematics`, it looks like the spec default is always true? Please inform me if otherwise / if needs to be changed

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
Make sure that if your custom collection needs `--spec` to be passed as default to the action, that you in the future write `--spec true` or add it to your `nest-cli.json` configuration file

Looking forward to comments and feedback :)
- Emil G

PS. I'm not sure where to add docs to this feature? If you tell me, I will try and write some :)

kamilmysliwiec commented:
Amazing contribution, thank you!

One thing before I dive into the code: would it be possible to avoid introducing breaking changes? I'm referring to this change:

"The generate command, has been changed such that the spec argument is no longer default to true, but undefined, unless otherwise specified in the CLI arguments."

Also, 2 questions:

1. Is `--no-spec` still supported?
2. Can we treat `--spec` (without `true`) same as `--spec true` (less typing, pretty much the same meaning)

> PS. I'm not sure where to add docs to this feature? If you tell me, I will try and write some :)

Here's the docs repo https://github.com/nestjs/docs.nestjs.com. We should update both CLI->Usage (https://docs.nestjs.com/cli/usages, `nest generate` options section) and CLI->Workspaces chapters (https://docs.nestjs.com/cli/monorepo, new section?)

egodsk commented:
Hi @kamilmysliwiec 

Thank you for your feedback!

I have looked back at my PR, and changed a couple of things related to your comment. First off to answer your questions:
> 1. Is --no-spec still supported?
Of course! you can either supply `--no-spec` or `--spec` to the CLI in order to overwrite any `generateOptions`

> 2. Can we treat --spec (without true) same as --spec true (less typing, pretty much the same meaning)
I have changed `--spec` to now support `--spec` *ONLY*, and no longer with either true or false (Use `--spec` for true and `--no-spec` for false).

I was not able to find a solution which allowed for both the shorthand `--spec` as well as `--spec [true/false]` (Believe me i tried), so I removed the ability to use `--spec [true/false]`, which kind of only also makes the cli cleaner in a way :) 

In regards to your comment about breaking changes:
> One thing before I dive into the code: would it be possible to avoid introducing breaking changes? I'm referring to this change:
>
> "The generate command, has been changed such that the spec argument is no longer default to true, but undefined, unless otherwise specified in the CLI arguments."

I have found and implemented a solution which now again supports the default of true, such that no breaking changes should be in this PR! Hurra!

The way I handle this, is by generating a custom object in the case of the `--spec` command being specifically specified in the CLI, such that we know when `spec` was specified in the CLI or if it was just assigned to the default value (Could not find a way to see via the Commander.js library, so implemented it this way instead).

egodsk commented:
Have added a PR on the docs repo: https://github.com/nestjs/docs.nestjs.com/pull/931

egodsk commented:
Quick rebase on nestjs/nest-cli to resolve conflicts :)

kamilmysliwiec commented on lib/configuration/configuration.ts:
```diff
@@ -15,6 +15,14 @@ interface CompilerOptions {
+    | boolean
+    | {
+        [key: string]: boolean;
+      };
```
What about using `Record<string, boolean>` instead?

egodsk replied:
I guess it could be used instead, however I don’t really see the real benefit of changing it to that instead? Could you explain?

kamilmysliwiec replied:
This type is predefined and comes with TypeScript (so there is no reason to define a new one) + we're using `Record` a lot in all Nest packages, hence, it would be nice to stay cohesive

egodsk replied:
Ah! Yes, will commit a change for this later.

kamilmysliwiec replied:
Thank you :)

kamilmysliwiec commented:
Thank you! Amazing contribution :)

John0x commented:
This has not yet been released on npm right?
Just making sure, cause it ain't working with nest cli 6.14.0

egodsk commented:
Hi @John0x

Do you mean there is a bug in the code, or that it does not seem like it has been released on npm?

abigmiu commented:
when i set `spec` as `false`,  but i use `nest g res` it still generate spec file 

![image](https://user-images.githubusercontent.com/43983920/192151705-9fce0e91-269e-434f-bbf9-e40e9c6a0bdd.png)

Haze272 commented:
This feature still doesn't work could we see it in nest updates?
